# <font color=green>1. MEU PRIMEIRO SCRAPING

# 1.1. Introdução

## *Web Scraping* é o termo utilizado para definir a prática de coletar automaticamente informações na Internet. Isto é feito, geralmente, por meio de programas que simulam a navegação humana na Web.

# 1.2. Ambiente e bibliotecas
### Utilizaremos em nosso treinamento o navegador Google Chrome

In [1]:
import bs4
import urllib.request as urllib_request
import pandas

print("BeautifulSoup ->", bs4.__version__)
print("urllib ->", urllib_request.__version__)
print("pandas ->", pandas.__version__)

BeautifulSoup -> 4.9.3
urllib -> 3.8
pandas -> 1.1.3


# 1.3. Meu primeiro scraping

In [2]:
from urllib.request import urlopen
from bs4 import BeautifulSoup

url = 'https://alura-site-scraping.herokuapp.com/hello-world.php'

response = urlopen(url)
html = response.read()

In [3]:
html

b'<!DOCTYPE html>\r\n<html lang="pt-br">\r\n<head>\r\n    <meta charset="utf-8">\r\n    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">\r\n\r\n    <title>Alura Motors</title>\r\n\t<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">\r\n\t<link rel="stylesheet" href="css/styles.css" media="all">\r\n\r\n\t<script src="https://code.jquery.com/jquery-1.12.4.js"></script>\r\n\t<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>\r\n\t<script type="text/javascript" src="js/index.js"></script>\r\n\r\n</head>\r\n<body cz-shortcut-listen="true">\r\n    <noscript>You need to enable JavaScript to run this app.</noscript>\r\n\r\n    <div id="root">\r\n        <h

In [4]:
soup = BeautifulSoup(html, 'html.parser')
soup

<!DOCTYPE html>

<html lang="pt-br">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1, shrink-to-fit=no" name="viewport"/>
<title>Alura Motors</title>
<link crossorigin="anonymous" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" rel="stylesheet"/>
<link href="css/styles.css" media="all" rel="stylesheet"/>
<script src="https://code.jquery.com/jquery-1.12.4.js"></script>
<script crossorigin="anonymous" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
<script src="js/index.js" type="text/javascript"></script>
</head>
<body cz-shortcut-listen="true">
<noscript>You need to enable JavaScript to run this app.</noscript>
<div id="root">
<header>
<nav class="navbar navbar-inverse" style="margin-bottom: 0;">
<div class="container" styl

Por exemplo, se quisermos acessar o elemento "Hello World!!!", clicaremos sobre este com o botão direito do mouse e escolheremos a opção "inspecionar". Em destaque, veremos que está dentro de uma tag **'h1'** com id igual a **'hello-world'**.

In [5]:
soup.find('h1', id='hello-world')

<h1 id="hello-world">Hello World!!!</h1>

Se quisermos acessar somente o conteúdo textual de "Hello World!!!” e não a tag inteira, poderemos usar o método .get_text() após .find() para retornarmos apenas o texto.

In [6]:
soup.find('h1', id='hello-world').get_text()

'Hello World!!!'

In [7]:
soup.find('p')

<p class="definition"><b>Web Scraping</b> é o termo utilizado para definir a prática de coletar automaticamente informações na Internet. Isto é feito, geralmente, por meio de programas que simulam a navegação humana na Web.</p>

In [8]:
soup.find('p').get_text()

'Web Scraping é o termo utilizado para definir a prática de coletar automaticamente informações na Internet. Isto é feito, geralmente, por meio de programas que simulam a navegação humana na Web.'

---
# <font color=green>2. OBTENDO E TRATANDO O CONTEÚDO DE UM HTML

# 2.1. Entendendo a web

<img src="./web/web.png" width="700">

# 2.2. Obtendo o conteúdo HTML de um site

[urllib.request](https://docs.python.org/3/library/urllib.html)

In [11]:
from urllib.request import urlopen

url = 'https://alura-site-scraping.herokuapp.com/index.php'

response = urlopen(url)
html = response.read()
html

b'<!DOCTYPE html>\r\n<html lang="pt-br">\r\n<head>\r\n    <meta charset="utf-8">\r\n    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">\r\n\r\n    <title>Alura Motors</title>\r\n\r\n\t<style>\r\n\t\t/*Regra para a animacao*/\r\n\t\t@keyframes spin {\r\n\t\t\t0% { transform: rotate(0deg); }\r\n\t\t\t100% { transform: rotate(360deg); }\r\n\t\t}\r\n\t\t/*Mudando o tamanho do icone de resposta*/\r\n\t\tdiv.glyphicon {\r\n\t\t\tcolor:#6B8E23;\r\n\t\t\tfont-size: 38px;\r\n\t\t}\r\n\t\t/*Classe que mostra a animacao \'spin\'*/\r\n\t\t.loader {\r\n\t\t\tborder: 16px solid #f3f3f3;\r\n\t\t\tborder-radius: 50%;\r\n\t\t\tborder-top: 16px solid #3498db;\r\n\t\t\twidth: 80px;\r\n\t\t\theight: 80px;\r\n\t\t\t-webkit-animation: spin 2s linear infinite;\r\n\t\t\tanimation: spin 2s linear infinite;\r\n\t\t}\r\n\t</style>\r\n\t<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuH

[urllib](https://docs.python.org/3/library/urllib.request.html#urllib.request.Request)

In [12]:
from urllib.request import Request, urlopen
from urllib.error import URLError, HTTPError

url = 'https://www.alura.com.br'
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.100 Safari/537.36'}



In [14]:
try:
    req = Request(url, headers = headers)
    response = urlopen(req)
    print(response.read())
    
except HTTPError as e:
    print(e.status, e.reason)
    
except URLError as e:
    print(e.reason)

b'\n\n\n\n<!DOCTYPE html><html\nlang="pt-BR"><head><meta\ncharset="UTF-8"><meta\nname="viewport" content="width=device-width,initial-scale=1,minimum-scale=1.0"><title>Alura | Cursos online de Tecnologia</title><meta\nname="description" content="Aprenda Programa\xc3\xa7\xc3\xa3o, Front-end, Data Science, UX, DevOps, Marketing, Inova\xc3\xa7\xc3\xa3o e Gest\xc3\xa3o na maior plataforma de tecnologia do Brasil"><link\nrel="canonical" href="https://www.alura.com.br"><link\nrel="icon" href="/assets/favicon.1621287442.ico" /><link\nhref="https://fonts.googleapis.com/css?display=swap&family=Inter:wght@400;700;900" rel="stylesheet" crossorigin><link\nrel="preconnect" href="https://fonts.gstatic.com/" crossorigin><link\nrel="stylesheet" href="/bundle,base/_reset,base/base,base/buttons,base/colors,base/titulos.1619635891.css"><link\nrel="stylesheet" href="/bundle,home/homeNova/career-colors,home/homeNova/careers,home/homeNova/companies,home/homeNova/features,home/homeNova/home,home/homeNova/mobi

# 2.3. Tratamento de string

In [15]:
type(html)

bytes

### Convertando o tipo bytes para string

In [16]:
html = html.decode('utf-8')

In [17]:
type(html)

str

In [18]:
html

'<!DOCTYPE html>\r\n<html lang="pt-br">\r\n<head>\r\n    <meta charset="utf-8">\r\n    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">\r\n\r\n    <title>Alura Motors</title>\r\n\r\n\t<style>\r\n\t\t/*Regra para a animacao*/\r\n\t\t@keyframes spin {\r\n\t\t\t0% { transform: rotate(0deg); }\r\n\t\t\t100% { transform: rotate(360deg); }\r\n\t\t}\r\n\t\t/*Mudando o tamanho do icone de resposta*/\r\n\t\tdiv.glyphicon {\r\n\t\t\tcolor:#6B8E23;\r\n\t\t\tfont-size: 38px;\r\n\t\t}\r\n\t\t/*Classe que mostra a animacao \'spin\'*/\r\n\t\t.loader {\r\n\t\t\tborder: 16px solid #f3f3f3;\r\n\t\t\tborder-radius: 50%;\r\n\t\t\tborder-top: 16px solid #3498db;\r\n\t\t\twidth: 80px;\r\n\t\t\theight: 80px;\r\n\t\t\t-webkit-animation: spin 2s linear infinite;\r\n\t\t\tanimation: spin 2s linear infinite;\r\n\t\t}\r\n\t</style>\r\n\t<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHA

### Eliminando os caracteres de tabulação, quebra de linha etc.

In [16]:
html.split()

['<!DOCTYPE',
 'html>',
 '<html',
 'lang="pt-br">',
 '<head>',
 '<meta',
 'charset="utf-8">',
 '<meta',
 'name="viewport"',
 'content="width=device-width,',
 'initial-scale=1,',
 'shrink-to-fit=no">',
 '<title>Alura',
 'Motors</title>',
 '<style>',
 '/*Regra',
 'para',
 'a',
 'animacao*/',
 '@keyframes',
 'spin',
 '{',
 '0%',
 '{',
 'transform:',
 'rotate(0deg);',
 '}',
 '100%',
 '{',
 'transform:',
 'rotate(360deg);',
 '}',
 '}',
 '/*Mudando',
 'o',
 'tamanho',
 'do',
 'icone',
 'de',
 'resposta*/',
 'div.glyphicon',
 '{',
 'color:#6B8E23;',
 'font-size:',
 '38px;',
 '}',
 '/*Classe',
 'que',
 'mostra',
 'a',
 'animacao',
 "'spin'*/",
 '.loader',
 '{',
 'border:',
 '16px',
 'solid',
 '#f3f3f3;',
 'border-radius:',
 '50%;',
 'border-top:',
 '16px',
 'solid',
 '#3498db;',
 'width:',
 '80px;',
 'height:',
 '80px;',
 '-webkit-animation:',
 'spin',
 '2s',
 'linear',
 'infinite;',
 'animation:',
 'spin',
 '2s',
 'linear',
 'infinite;',
 '}',
 '</style>',
 '<link',
 'rel="stylesheet"',
 'hre

In [17]:
" ".join(html.split())

'<!DOCTYPE html> <html lang="pt-br"> <head> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <title>Alura Motors</title> <style> /*Regra para a animacao*/ @keyframes spin { 0% { transform: rotate(0deg); } 100% { transform: rotate(360deg); } } /*Mudando o tamanho do icone de resposta*/ div.glyphicon { color:#6B8E23; font-size: 38px; } /*Classe que mostra a animacao \'spin\'*/ .loader { border: 16px solid #f3f3f3; border-radius: 50%; border-top: 16px solid #3498db; width: 80px; height: 80px; -webkit-animation: spin 2s linear infinite; animation: spin 2s linear infinite; } </style> <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"> <link rel="stylesheet" href="css/styles.css" media="all"> <script src="https://code.jquery.com/jquery-1.12.4.js"></script> <script src="https://

### Eliminando os espaços em branco entre as TAGS

In [18]:
" ".join(html.split()).replace('> <', '><')

'<!DOCTYPE html><html lang="pt-br"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Alura Motors</title><style> /*Regra para a animacao*/ @keyframes spin { 0% { transform: rotate(0deg); } 100% { transform: rotate(360deg); } } /*Mudando o tamanho do icone de resposta*/ div.glyphicon { color:#6B8E23; font-size: 38px; } /*Classe que mostra a animacao \'spin\'*/ .loader { border: 16px solid #f3f3f3; border-radius: 50%; border-top: 16px solid #3498db; width: 80px; height: 80px; -webkit-animation: spin 2s linear infinite; animation: spin 2s linear infinite; } </style><link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><link rel="stylesheet" href="css/styles.css" media="all"><script src="https://code.jquery.com/jquery-1.12.4.js"></script><script src="https://maxcdn.boo

### Função de tratamento de strings

In [19]:
def trata_html(input):
    return " ".join(input.split()).replace('> <', '><')

In [20]:
html = trata_html(html)

In [21]:
html

'<!DOCTYPE html><html lang="pt-br"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Alura Motors</title><style> /*Regra para a animacao*/ @keyframes spin { 0% { transform: rotate(0deg); } 100% { transform: rotate(360deg); } } /*Mudando o tamanho do icone de resposta*/ div.glyphicon { color:#6B8E23; font-size: 38px; } /*Classe que mostra a animacao \'spin\'*/ .loader { border: 16px solid #f3f3f3; border-radius: 50%; border-top: 16px solid #3498db; width: 80px; height: 80px; -webkit-animation: spin 2s linear infinite; animation: spin 2s linear infinite; } </style><link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><link rel="stylesheet" href="css/styles.css" media="all"><script src="https://code.jquery.com/jquery-1.12.4.js"></script><script src="https://maxcdn.boo

---
# <font color=green>3. INTRODUÇÃO AO BEAUTIFULSOUP

# 3.1. HTML da nossa página

**HTML** (*HyperText Markup Language*) é uma linguagem de marcação composta por **tags** que deteminam o papel que cada parte do documento vai assumir. As **tags** são formadas pelo seu nome e atributos. Os atributos servem para configurar e também modificar as características padrões de uma **tag**.

## Estrutura básica

```html
<html>
    <head>
        <meta charset="utf-8" />
        <title>Alura Motors</title>
    </head>
    <body>
        <div id="container">
            <h1>Alura</h1>
            <h2 class="formato">Cursos de Tecnologia</h2>
            <p>Você vai estudar, praticar, discutir e aprender.</p>
            <a href="https://www.alura.com.br/">Clique aqui</a>
        </div>
    </body>
</html>
```

```<html>``` - determina o início do documento.

```<head>``` - cabeçalho. Contém informações e configurações do documento.

```<body>``` - é o corpo do documento, onde todo o conteúdo é colocado. Esta é a parte visível em um navegador.

## Tags mais comuns

```<div>``` - define uma divisão da página. Pode ser formatada de diversas maneiras.

```<h1>, <h2>, <h3>, <h4>, <h5>, <h6>``` - marcadores de títulos.

```<p>``` - marcador de parágrafo.

```<a>``` - hiperlink.

```<img>``` - exibição de imagens.

```<table>``` - definição de tabelas.

```<ul>, <li>``` - definição de listas.


## A Estrutura básica acima diretamente no Markdown

<html>
    <head>
        <meta charset="utf-8" />
        <title>Alura Motors</title>
    </head>
    <body>
        <div id="container">
            <h1>Alura</h1>
            <h2 class="formato">Cursos de Tecnologia</h2>
            <p>Você vai estudar, praticar, discutir e aprender.</p>
            <a href="https://www.alura.com.br/">Clique aqui</a>
        </div>
    </body>
</html>

# 3.2. Criando um objeto BeautifulSoup

[BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/)

Sobre parser [aqui](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#parser-installation)


In [24]:
from bs4 import BeautifulSoup

soup = BeautifulSoup(html, 'html.parser')

soup

<!DOCTYPE html>
<html lang="pt-br"><head><meta charset="utf-8"/><meta content="width=device-width, initial-scale=1, shrink-to-fit=no" name="viewport"/><title>Alura Motors</title><style> /*Regra para a animacao*/ @keyframes spin { 0% { transform: rotate(0deg); } 100% { transform: rotate(360deg); } } /*Mudando o tamanho do icone de resposta*/ div.glyphicon { color:#6B8E23; font-size: 38px; } /*Classe que mostra a animacao 'spin'*/ .loader { border: 16px solid #f3f3f3; border-radius: 50%; border-top: 16px solid #3498db; width: 80px; height: 80px; -webkit-animation: spin 2s linear infinite; animation: spin 2s linear infinite; } </style><link crossorigin="anonymous" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" rel="stylesheet"/><link href="css/styles.css" media="all" rel="stylesheet"/><script src="https://code.jquery.com/jquery-1.12.4.js"></script><script crossorigin="anonymou

Como segundo parâmetro, adicionaremos o **'html.parser'** entre aspas simples. Conseguiremos executar da mesma forma, mas se não o colocarmos, veremos uma mensagem indicando que poderemos ter resultados diferentes caso mudemos de computador, e que com isso usará o mais correto automaticamente.

Uma forma de mostrá-lo de maneira mais elegante, já que fizemos o tratamento, é por meio de **print()** recebendo **soup.prettify**, que nos retornará o texto com formato mais parecido com uma árvore já com todas as informações.

In [25]:
print(soup.prettify())

<!DOCTYPE html>
<html lang="pt-br">
 <head>
  <meta charset="utf-8"/>
  <meta content="width=device-width, initial-scale=1, shrink-to-fit=no" name="viewport"/>
  <title>
   Alura Motors
  </title>
  <style>
   /*Regra para a animacao*/ @keyframes spin { 0% { transform: rotate(0deg); } 100% { transform: rotate(360deg); } } /*Mudando o tamanho do icone de resposta*/ div.glyphicon { color:#6B8E23; font-size: 38px; } /*Classe que mostra a animacao 'spin'*/ .loader { border: 16px solid #f3f3f3; border-radius: 50%; border-top: 16px solid #3498db; width: 80px; height: 80px; -webkit-animation: spin 2s linear infinite; animation: spin 2s linear infinite; }
  </style>
  <link crossorigin="anonymous" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" rel="stylesheet"/>
  <link href="css/styles.css" media="all" rel="stylesheet"/>
  <script src="https://code.jquery.com/jquery-1.12.4.js">
  

# 3.3. Acessando tags

In [26]:
soup.html

<html lang="pt-br"><head><meta charset="utf-8"/><meta content="width=device-width, initial-scale=1, shrink-to-fit=no" name="viewport"/><title>Alura Motors</title><style> /*Regra para a animacao*/ @keyframes spin { 0% { transform: rotate(0deg); } 100% { transform: rotate(360deg); } } /*Mudando o tamanho do icone de resposta*/ div.glyphicon { color:#6B8E23; font-size: 38px; } /*Classe que mostra a animacao 'spin'*/ .loader { border: 16px solid #f3f3f3; border-radius: 50%; border-top: 16px solid #3498db; width: 80px; height: 80px; -webkit-animation: spin 2s linear infinite; animation: spin 2s linear infinite; } </style><link crossorigin="anonymous" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" rel="stylesheet"/><link href="css/styles.css" media="all" rel="stylesheet"/><script src="https://code.jquery.com/jquery-1.12.4.js"></script><script crossorigin="anonymous" integrity="sh

In [27]:
# acessar apenas o título da página do <head>
# Para chegarmos no <title>, escreveremos apenas .title ao final do código anterior

soup.html.head.title

<title>Alura Motors</title>

In [28]:
soup.title

<title>Alura Motors</title>

In [29]:
soup.div.div.div.div.h5

<h5 class="modal-title" id="loadingModal_label"><span class="glyphicon glyphicon-refresh"></span>Aguarde... </h5>

In [30]:
soup.h5

<h5 class="modal-title" id="loadingModal_label"><span class="glyphicon glyphicon-refresh"></span>Aguarde... </h5>

# 3.4. Acessando o conteúdo das tags

In [31]:
soup.html.head.title

<title>Alura Motors</title>

In [32]:
soup.title

<title>Alura Motors</title>

In [33]:
soup.title.get_text()

'Alura Motors'

In [34]:
soup.h5.getText() # outra forma de usar o get_text()

'Aguarde... '

In [35]:
soup.getText()

'Alura MotorsYou need to enable JavaScript to run this app.Aguarde... Motors Hello WorldAnúnciosVeículos de Luxo Novos e Usados - Todas as Marcas246 veículos encontradosPágina 1 de 25LAMBORGHINI AVENTADORUSADOMotor 1.8 16vAno 1993 - 55.286 km► 4 X 4► Câmera de estacionamento► Controle de tração► Sensor de estacionamento...Belo Horizonte - MGR$ 338.000BMW M2USADOMotor 3.0 32vAno 2018 - 83.447 km► Câmera de estacionamento► Controle de estabilidade► Travas elétricas► Freios ABS...Belo Horizonte - MGR$ 346.000ALFAUSADOMotor 1.8 16vAno 2004 - 19.722 km► Central multimídia► Bancos de couro► Rodas de liga► Câmera de estacionamento...Rio de Janeiro - RJR$ 480.000PUECHUSADOMotor Diesel V8Ano 1992 - 34.335 km► Bancos de couro► Freios ABS► Rodas de liga► Câmbio automático...São Paulo - SPR$ 133.000LAMBORGHINI MURCIELAGOUSADOMotor 1.0 8vAno 1991 - 464 km► Central multimídia► Teto panorâmico► Sensor crepuscular► Piloto automático...Belo Horizonte - MGR$ 175.000ASTON MARTINUSADOOPORTUNIDADEMotor Die

# 3.5. Acessando os atributos de uma tag

Escreveremos soup.img para retornarmos a primeira tag encontrada com este nome, bem como todos seus atributos.

In [36]:
soup.img

<img alt="Alura" class="d-inline-block align-top" src="img/alura-logo.svg"/>

Para os acessarmos de uma só vez, usaremos .attrs.

In [37]:
soup.img.attrs

{'src': 'img/alura-logo.svg',
 'class': ['d-inline-block', 'align-top'],
 'alt': 'Alura'}

Se quisermos acessar somente as chaves e saber quais são os atributos da minha tag, adicionaremos o método **.keys()** para recebermos um dicionário com as chaves.

In [38]:
soup.img.attrs.keys()

dict_keys(['src', 'class', 'alt'])

Se quisermos os valores, usaremos .values() para exibirmos um dicionário com os valores.

In [40]:
soup.img.attrs.values()

dict_values(['img/alura-logo.svg', ['d-inline-block', 'align-top'], 'Alura'])

Outra forma de acessarmos os atributos de uma tag que podem facilitar o nosso trabalho é escrever soup.img[] com chaves na célula seguinte, como se estivéssemos acessando um dicionário.

Ou seja, se quisermos acessar somente a informação de classe no dicionário, passaremos 'class' entre aspas simples.

In [41]:
soup.img['class']

['d-inline-block', 'align-top']

Como retorno, veremos uma lista com as informações deste atributo em especial.

Um outro jeito de fazer será escrevermos soup.img.get() na célula seguinte passando o nome do atributo que queremos acessar, neste caso o 'src'.

In [42]:
soup.img.get('src')

'img/alura-logo.svg'

---
# <font color=green>4. PESQUISANDO COM O BEAUTIFULSOUP

# 4.1. Os métodos *find()* e *findAll()*

- ### *find(tag, attributes, recursive, text, **kwargs)*

- ### *findAll(tag, attributes, recursive, text, limit, **kwargs)*

#### https://www.crummy.com/software/BeautifulSoup/bs4/doc/#find
#### https://www.crummy.com/software/BeautifulSoup/bs4/doc/#find-all

> **Observação:**
> - *findAll()* também pode ser utilizado como *find_all()*

### Método *find()*

Na primeira célula desta parte, escreveremos apenas soup.find() recebendo o 'img' para encontrarmos esta primeira tag, como vimos anteriormente com soup.img.

In [43]:
soup.find('img')

<img alt="Alura" class="d-inline-block align-top" src="img/alura-logo.svg"/>

In [44]:
soup.img

<img alt="Alura" class="d-inline-block align-top" src="img/alura-logo.svg"/>

### Método *findAll()*

Já o findAll() é mais interessante, pois poderá encontrar todas as tags que estamos procurando, da mesma forma que fizemos com a busca no inspetor de código no Chrome onde apareceram 12 ocorrências de "img" quando buscamos pela marcação na barra de pesquisa.

In [45]:
soup.findAll('img')

[<img alt="Alura" class="d-inline-block align-top" src="img/alura-logo.svg"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/bmw-m2/bmw-m2-2970882__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/alfa/alfa-1823056__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/puech/puech-4055386__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-murcielago/lamborghini-murcielago-2872974__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/

### Comando equivalente ao método *find()*

In [46]:
soup.findAll('img', limit =1)[0]

<img alt="Alura" class="d-inline-block align-top" src="img/alura-logo.svg"/>

### Atalho para o método *findAll()*

In [48]:
soup('img')

[<img alt="Alura" class="d-inline-block align-top" src="img/alura-logo.svg"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/bmw-m2/bmw-m2-2970882__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/alfa/alfa-1823056__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/puech/puech-4055386__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-murcielago/lamborghini-murcielago-2872974__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/

### Passando listas de TAGs

In [49]:
soup.findAll(['h1', 'h2', 'h3', 'h4', 'h5', 'h6'])

[<h5 class="modal-title" id="loadingModal_label"><span class="glyphicon glyphicon-refresh"></span>Aguarde... </h5>,
 <h4><b id="loadingModal_content"></b></h4>,
 <h1 class="sub-header">Veículos de Luxo Novos e Usados - Todas as Marcas</h1>]

### Utilizando o argumento *attributes*

In [54]:
soup.findAll('p', {'class': 'txt-value'})

[<p class="txt-value">R$ 338.000</p>,
 <p class="txt-value">R$ 346.000</p>,
 <p class="txt-value">R$ 480.000</p>,
 <p class="txt-value">R$ 133.000</p>,
 <p class="txt-value">R$ 175.000</p>,
 <p class="txt-value">R$ 239.000</p>,
 <p class="txt-value">R$ 115.000</p>,
 <p class="txt-value">R$ 114.000</p>,
 <p class="txt-value">R$ 75.000</p>,
 <p class="txt-value">R$ 117.000</p>]

### Buscando por conteúdo de uma TAG

In [56]:
soup.findAll('p', text = 'Belo Horizonte - MG')

[<p class="txt-location">Belo Horizonte - MG</p>,
 <p class="txt-location">Belo Horizonte - MG</p>,
 <p class="txt-location">Belo Horizonte - MG</p>,
 <p class="txt-location">Belo Horizonte - MG</p>,
 <p class="txt-location">Belo Horizonte - MG</p>]

### Utilizando diretamente os atributos

In [57]:
soup.findAll('img', alt='Foto')

[<img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/bmw-m2/bmw-m2-2970882__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/alfa/alfa-1823056__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/puech/puech-4055386__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-murcielago/lamborghini-murcielago-2872974__340.jpg" width="220"/>,
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/aston-martin/aston-martin-2977916__340.jpg" width="220"/>,
 <img al

In [60]:
for item in soup.findAll('img', alt='Foto'):
    print(item.get('src'))

https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg
https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/bmw-m2/bmw-m2-2970882__340.jpg
https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/alfa/alfa-1823056__340.jpg
https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/puech/puech-4055386__340.jpg
https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-murcielago/lamborghini-murcielago-2872974__340.jpg
https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/aston-martin/aston-martin-2977916__340.jpg
https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/tvr/tvr-2943925__340.jpg
https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/excalibur/excalibur-2916730__340.jpg
https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/mclaren/mclaren-2855240__340.jpg
htt

### Cuidado com o atributo "class"

In [63]:
soup.findAll('p', class_='txt-value')    # class é palavra reservada do python, para funcionar, coloque um _ depois

[<p class="txt-value">R$ 338.000</p>,
 <p class="txt-value">R$ 346.000</p>,
 <p class="txt-value">R$ 480.000</p>,
 <p class="txt-value">R$ 133.000</p>,
 <p class="txt-value">R$ 175.000</p>,
 <p class="txt-value">R$ 239.000</p>,
 <p class="txt-value">R$ 115.000</p>,
 <p class="txt-value">R$ 114.000</p>,
 <p class="txt-value">R$ 75.000</p>,
 <p class="txt-value">R$ 117.000</p>]

### Obtendo todo o conteúdo de texto de uma página

In [64]:
soup.findAll(text = True)

['html',
 'Alura Motors',
 " /*Regra para a animacao*/ @keyframes spin { 0% { transform: rotate(0deg); } 100% { transform: rotate(360deg); } } /*Mudando o tamanho do icone de resposta*/ div.glyphicon { color:#6B8E23; font-size: 38px; } /*Classe que mostra a animacao 'spin'*/ .loader { border: 16px solid #f3f3f3; border-radius: 50%; border-top: 16px solid #3498db; width: 80px; height: 80px; -webkit-animation: spin 2s linear infinite; animation: spin 2s linear infinite; } ",
 ' requisitar("https://alura-scraping.herokuapp.com/produtos?_page="+1+"&_limit=10") ',
 'You need to enable JavaScript to run this app.',
 'Aguarde... ',
 'Motors ',
 'Hello World',
 'Anúncios',
 'Veículos de Luxo Novos e Usados - Todas as Marcas',
 '246 veículos encontrados',
 'Página 1 de 25',
 'LAMBORGHINI AVENTADOR',
 'USADO',
 'Motor 1.8 16v',
 'Ano 1993 - 55.286 km',
 '► 4 X 4',
 '► Câmera de estacionamento',
 '► Controle de tração',
 '► Sensor de estacionamento',
 '...',
 'Belo Horizonte - MG',
 'R$ 338.000',

# 4.2. Outros métodos de pesquisa

- ### *findParent(tag, attributes, text, **kwargs)*

- ### *findParents(tag, attributes, text, limit, **kwargs)*

#### https://www.crummy.com/software/BeautifulSoup/bs4/doc/#find-parents-and-find-parent

> **Observação:**
> - *findParent()* e *findParents()* também podem ser utilizados como *find_parent()* e *find_parents()*, respectivamente.
---
- ### *findNextSibling(tag, attributes, text, **kwargs)*

- ### *findNextSiblings(tag, attributes, text, limit, **kwargs)*

- ### *findPreviousSibling(tag, attributes, text, **kwargs)*

- ### *findPreviousSiblings(tag, attributes, text, limit, **kwargs)*

#### https://www.crummy.com/software/BeautifulSoup/bs4/doc/#find-next-siblings-and-find-next-sibling
#### https://www.crummy.com/software/BeautifulSoup/bs4/doc/#find-previous-siblings-and-find-previous-sibling

> **Observação:**
> - *findNextSibling()*, *findNextSiblings()*, *findPreviousSibling()* e *findPreviousSiblings()* também podem ser utilizados como *find_next_sibling()*, *find_next_siblings()*, *find_previous_sibling()* e *find_previous_siblings()*, respectivamente.
---
- ### *findNext(tag, attributes, text, **kwargs)*

- ### *findAllNext(tag, attributes, text, limit, **kwargs)*

- ### *findPrevious(tag, attributes, text, **kwargs)*

- ### *findAllPrevious(tag, attributes, text, limit, **kwargs)*

#### https://www.crummy.com/software/BeautifulSoup/bs4/doc/#find-all-next-and-find-next
#### https://www.crummy.com/software/BeautifulSoup/bs4/doc/#find-all-previous-and-find-previous

> **Observação:**
> - *findNext()*, *findAllNext()*, *findPrevious* e *findAllPrevious* também podem ser utilizados como *find_next()*, *find_all_next()*, *find_previous()* e *find_all_previous()*, respectivamente.

## HTML de exemplo para ilustrar a utilização dos métodos de pesquisa do BeautifulSoup

<img src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/BeautifulSoup-method.png" width=80%>

---
## Resultado

<html>
    <body>
        <div id=“container-a”>
            <h1>Título A</h1>
            <h2 class="ref-a">Sub título A</h2>
            <p>Texto de conteúdo A</p>
        </div>
        <div id=“container-b”>
            <h1>Título B</h1>
            <h2 class="ref-b">Sub título B</h2>
            <p>Texto de conteúdo B</p>
        </div>
    </body>
</html>

In [67]:
html_teste = """
    <html>
        <body>
            <div id="container-a">
                <h1>Título A</h1>
                <h2 class="ref-a">Sub título A</h2>
                <p>Texto de conteúdo A</p>
            </div>
            <div id="container-b">
                <h1>Título B</h1>
                <h2 class="ref-b">Sub título B</h2>
                <p>Texto de conteúdo B</p>
            </div>
        </body>
    </html>
"""

### Tratamentos para a string HTML

In [69]:
html_teste = trata_html(html_teste)

In [70]:
html_teste

'<html><body><div id="container-a"><h1>Título A</h1><h2 class="ref-a">Sub título A</h2><p>Texto de conteúdo A</p></div><div id="container-b"><h1>Título B</h1><h2 class="ref-b">Sub título B</h2><p>Texto de conteúdo B</p></div></body></html>'

### Criando o objeto BeautifulSoup

In [71]:
soup = BeautifulSoup(html_teste, 'html.parser')
soup

<html><body><div id="container-a"><h1>Título A</h1><h2 class="ref-a">Sub título A</h2><p>Texto de conteúdo A</p></div><div id="container-b"><h1>Título B</h1><h2 class="ref-b">Sub título B</h2><p>Texto de conteúdo B</p></div></body></html>

### Parents

Primeiro, usaremos o .find() recebendo 'h2'.

In [ ]:
soup.find('h2')

O método encontrará a primeira tag **h2** da cascata do arquivo HTML como já sabemos, mas em nosso exemplo há duas.

O BeautifulSoup nos permite fazer vários alinhamentos de métodos; se quisermos buscar o pai desta marcação, usaremos o .find_parent() recebendo 'div' neste caso, mas poderiam ser outros.

In [72]:
soup.find('h2').find_parent('div')

<div id="container-a"><h1>Título A</h1><h2 class="ref-a">Sub título A</h2><p>Texto de conteúdo A</p></div>

Já o .find_parents() nos reportará todos os pais acima do 'h2'.

In [73]:
soup.find('h2').find_parents()

[<div id="container-a"><h1>Título A</h1><h2 class="ref-a">Sub título A</h2><p>Texto de conteúdo A</p></div>,
 <body><div id="container-a"><h1>Título A</h1><h2 class="ref-a">Sub título A</h2><p>Texto de conteúdo A</p></div><div id="container-b"><h1>Título B</h1><h2 class="ref-b">Sub título B</h2><p>Texto de conteúdo B</p></div></body>,
 <html><body><div id="container-a"><h1>Título A</h1><h2 class="ref-a">Sub título A</h2><p>Texto de conteúdo A</p></div><div id="container-b"><h1>Título B</h1><h2 class="ref-b">Sub título B</h2><p>Texto de conteúdo B</p></div></body></html>,
 <html><body><div id="container-a"><h1>Título A</h1><h2 class="ref-a">Sub título A</h2><p>Texto de conteúdo A</p></div><div id="container-b"><h1>Título B</h1><h2 class="ref-b">Sub título B</h2><p>Texto de conteúdo B</p></div></body></html>]

Se usarmos o .findAll() recebendo 'h2', exibiremos as duas tags **h2** presentes no código de exemplo.
    
Porém, se adicionarmos o método .find_parent() recebendo 'div' na mesma linha, retornaremos uma mensagem de erro.    

In [76]:
soup.findAll('h2')

[<h2 class="ref-a">Sub título A</h2>, <h2 class="ref-b">Sub título B</h2>]

In [77]:
for item in soup.findAll('h2'):
    print(item.find_parent('div'))

<div id="container-a"><h1>Título A</h1><h2 class="ref-a">Sub título A</h2><p>Texto de conteúdo A</p></div>
<div id="container-b"><h1>Título B</h1><h2 class="ref-b">Sub título B</h2><p>Texto de conteúdo B</p></div>


## Siblings

Desta forma, encontrará as duas marcações <div> presentes no HTML.

Agora, iremos aos siblings que são os irmãos no mesmo patamar hierárquico da tag buscada.

Com o método **.findNextSibling()**, encontraremos a marcação irmã imediatamente após a pesquisada.

In [78]:
soup.find('h2').findNextSibling()

<p>Texto de conteúdo A</p>

Como retorno, receberemos o **p**, ou seja, a tag seguinte à **h2** dentro de **div**.

Já o anterior poderemos encontrar com o método **.findPreviousSibling()**.

In [79]:
soup.find('h2').findPreviousSibling()

<h1>Título A</h1>

Como resultado, receberemos **h1**, que é o elemento imediatamente anterior ao **h2**.

Agora, pegaremos os irmãos anteriores à marcação **p** com **.findPreviousSiblings()**.

In [81]:
soup.find('p').findPreviousSiblings()

[<h2 class="ref-a">Sub título A</h2>, <h1>Título A</h1>]

## Next e Previous

Com isso, retornaremos as tags **h2** e **h1**.

Estes métodos sem comportam da mesma forma que o **.find_parent()**, varrendo de um ponto específico até o topo do texto.

Para buscarmos o elemento seguinte ao **h2**, usaremos o método **.findNext()** recebendo **'h2'**.

In [82]:
soup.find('h2').findNext()

<p>Texto de conteúdo A</p>

Obviamente, receberemos **p** como resultado da execução.

Para termos o anterior, usaremos **.findPrevious()**.

In [83]:
soup.find('h2').findPrevious()

<h1>Título A</h1>

Como retorno, teremos o **h1**.

Se quisermos todos os elementos seguintes à primeira marcação usaremos o método **.findAllNext()**.

In [84]:
soup.find('h2').findAllNext()

[<p>Texto de conteúdo A</p>,
 <div id="container-b"><h1>Título B</h1><h2 class="ref-b">Sub título B</h2><p>Texto de conteúdo B</p></div>,
 <h1>Título B</h1>,
 <h2 class="ref-b">Sub título B</h2>,
 <p>Texto de conteúdo B</p>]

Como resultado da execução, teremos os conteúdos de **p**, **div**, **h1**, **h2** e o último **p**.


# <font color=green>5. WEB SCRAPING DO SITE ALURA MOTORS - OBTENDO OS DADOS DE UM ANÚNCIO

# 5.1. Identificando e selecionando os dados no HTML

### Obtendo o HTML e criando o objeto BeautifulSoup

In [85]:
response = urlopen('https://alura-site-scraping.herokuapp.com/index.php')
html = response.read().decode('utf-8')
soup = BeautifulSoup(html, 'html.parser')
soup

<!DOCTYPE html>

<html lang="pt-br">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1, shrink-to-fit=no" name="viewport"/>
<title>Alura Motors</title>
<style>
		/*Regra para a animacao*/
		@keyframes spin {
			0% { transform: rotate(0deg); }
			100% { transform: rotate(360deg); }
		}
		/*Mudando o tamanho do icone de resposta*/
		div.glyphicon {
			color:#6B8E23;
			font-size: 38px;
		}
		/*Classe que mostra a animacao 'spin'*/
		.loader {
			border: 16px solid #f3f3f3;
			border-radius: 50%;
			border-top: 16px solid #3498db;
			width: 80px;
			height: 80px;
			-webkit-animation: spin 2s linear infinite;
			animation: spin 2s linear infinite;
		}
	</style>
<link crossorigin="anonymous" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" rel="stylesheet"/>
<link href="css/styles.css" media="all" rel="stylesheet"/>
<script src="https://code.jquery

### Criando variávels para armazenar informações

In [86]:
cards = []
card = {}

### Obtendo os dados do primeiro CARD

In [87]:
anuncio = soup.find('div', {'class': 'well card'})
anuncio

<div class="well card">
<div class="col-md-3 image-card">
<img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg" width="220"/>
</div>
<div class="col-md-6 body-card">
<p class="txt-name inline">LAMBORGHINI AVENTADOR</p>
<p class="txt-category badge badge-secondary inline">USADO</p>
<p class="txt-motor">Motor 1.8 16v</p>
<p class="txt-description">Ano 1993 - 55.286 km</p>
<ul class="lst-items">
<li class="txt-items">► 4 X 4</li>
<li class="txt-items">► Câmera de estacionamento</li>
<li class="txt-items">► Controle de tração</li>
<li class="txt-items">► Sensor de estacionamento</li>
<li class="txt-items">...</li>
</ul>
<p class="txt-location">Belo Horizonte - MG</p>
</div>
<div class="col-md-3 value-card">
<div class="value">
<p class="txt-value">R$ 338.000</p>
</div>
</div>
</div>

# 5.2. Obtendo o VALOR do veículo anunciado

Já identificamos e selecionamos onde estão os dados do nosso primeiro card. Faremos um trabalho mais elaborado neste e depois replicaremos para os outros.

Começaremos a identificar pontualmente onde está cada uma informação para fazermos as coletas.

No inspetor de código em nossa página, veremos que o primeiro <div> de classes well e card possui outras três divisões internas relativas ao nome do veículo, demais características e valores.

Iniciaremos com este último que é mais simples. Também poderemos apenas clicar com o botão direito sobre o valor que queremos pegar e selecionar a opção "Inspecionar..." para encontrarmos a linha exata desta informação.

Este está dentro de uma <div> com class col-md-3 e value-card.

De volta ao nosso notebook no Colab, chegaremos na parte "5.2". Na primeira célula, chamaremos a variável anuncio criada anteriormente.

Selecionaremos essa <div> e varreremos dentro deste card. Em seguida, aplicaremos o .find() recebendo 'div' que possui 'class' com duas formatações, como já vimos.

Como estamos trabalhando dentro deste recorte e não no HTML inteiro, não precisaremos ser tão específicos.

In [88]:
anuncio.find('div', {'class': 'value-card'})

<div class="col-md-3 value-card">
<div class="value">
<p class="txt-value">R$ 338.000</p>
</div>
</div>

Executando a célula, exibiremos apenas o conteúdo da <div> selecionada com seus filhos.

Dentro, encontraremos uma tag **p** com class igual a txt-value e o valor do carro propriamente dito, e notaremos que não há nenhuma outra com esta mesma classe dentro de anuncio.

Para acessarmos o conteúdo de forma mais simples, copiaremos e colaremos o código anterior para alterarmos o argumento de .find() para 'p' e classe 'txt-value'.

Em seguida, pegaremos a informação textual com .getText() ao final da linha, como já sabemos.

In [89]:
anuncio.find('p', {'class': 'txt-value'}).getText()

'R$ 338.000'

Com isso, exibiremos 'R$ 338.000' como resultado e teremos realizado nossa primeira raspagem de dados.

A variável card que é um dicionário Python será chamada em seguida; dentro dos colchetes, chamaremos a chave 'value' sendo igual ao código que exibiu o valor do veículo.

In [90]:
card['value'] = anuncio.find('p', {'class': 'txt-value'}).getText()

In [92]:
card

{'value': 'R$ 338.000'}

Rodando este código e chamando apenas card na célula seguinte, teremos a primeira informação dentro do dicionário.

Ao final, pegaremos todos os pontos e os juntaremos para selecionarmos os dados, e depois faremos a iteração em todos na página e depois no site.

A seguir, pegaremos outras partes dos dados do card.

### <font color=red>Resumo

In [93]:
# Valor
card['value'] = anuncio.find('p', {'class': 'txt-value'}).getText()

# 5.3. Obtendo as INFORMAÇÕES sobre o veículo anunciado

Agora que já conseguimos obter o valor do nosso card, começaremos a trabalhar com as informações do veículo, seu nome, motorização, acessórios e etc.

Dentro da **div** com class igual a "well card", teremos a classe do meio com body-card contendo as várias tags <p> que pegaremos com os dados do primeiro carro.

Na parte "5.3" de nosso notebook, chamaremos o anuncio com .find() recebendo a 'div' e a classe 'body-card'.

In [94]:
anuncio.find('div', {'class': 'body-card'})

<div class="col-md-6 body-card">
<p class="txt-name inline">LAMBORGHINI AVENTADOR</p>
<p class="txt-category badge badge-secondary inline">USADO</p>
<p class="txt-motor">Motor 1.8 16v</p>
<p class="txt-description">Ano 1993 - 55.286 km</p>
<ul class="lst-items">
<li class="txt-items">► 4 X 4</li>
<li class="txt-items">► Câmera de estacionamento</li>
<li class="txt-items">► Controle de tração</li>
<li class="txt-items">► Sensor de estacionamento</li>
<li class="txt-items">...</li>
</ul>
<p class="txt-location">Belo Horizonte - MG</p>
</div>

Como resultado da execução, veremos as informações dessa divisão com várias tags **p**, bem como **ul** e **li** que veremos adiante.

Para selecionarmos apenas as marcações **p**, usaremos o .findAll() recebendo apenas 'p' na célula seguinte.

In [95]:
anuncio.find('div', {'class': 'body-card'}).findAll('p')

[<p class="txt-name inline">LAMBORGHINI AVENTADOR</p>,
 <p class="txt-category badge badge-secondary inline">USADO</p>,
 <p class="txt-motor">Motor 1.8 16v</p>,
 <p class="txt-description">Ano 1993 - 55.286 km</p>,
 <p class="txt-location">Belo Horizonte - MG</p>]

Ao executarmos, receberemos uma lista apenas com essas tags, sendo cada uma com class diferentes: txt-name, txt-category, txt-motor, txt-description e txt-location.

Como explicitam do que se tratam, usaremos essas classes para nomearmos as variáveis e pegarmos as informações.

Na célula seguinte, armazenaremos a busca em infos sendo igual ao último comando, pois iremos iterar em cima desta lista recém criada.

In [96]:
infos = anuncio.find('div', {'class': 'body-card'}).findAll('p')

In [97]:
infos

[<p class="txt-name inline">LAMBORGHINI AVENTADOR</p>,
 <p class="txt-category badge badge-secondary inline">USADO</p>,
 <p class="txt-motor">Motor 1.8 16v</p>,
 <p class="txt-description">Ano 1993 - 55.286 km</p>,
 <p class="txt-location">Belo Horizonte - MG</p>]

Em seguida, aplicaremos for para cada info dentro de infos. Faremos um print() recebendo info para vermos o que possui.

In [98]:
for info in infos:
    print(info)

<p class="txt-name inline">LAMBORGHINI AVENTADOR</p>
<p class="txt-category badge badge-secondary inline">USADO</p>
<p class="txt-motor">Motor 1.8 16v</p>
<p class="txt-description">Ano 1993 - 55.286 km</p>
<p class="txt-location">Belo Horizonte - MG</p>


Logo, veremos cada item da lista para fazermos a iteração dentro desse caso.

Como queremos os nomes da class para os jogarmos dentro da variável card, usaremos os mesmos já presentes em cada classe de **p** encontrada.

Para pegarmos o valor de um atributo de uma tag HTML, usaremos .get() recebendo 'class' em info do print().

In [99]:
for info in infos:
    print(info.get('class'))

['txt-name', 'inline']
['txt-category', 'badge', 'badge-secondary', 'inline']
['txt-motor']
['txt-description']
['txt-location']


Como retorno, recebemos os atributos das classes. Sempre no primeiro item da lista retornada, teremos a informação que nos interessa.

Para facilitar nosso trabalho, adicionaremos apenas um traço ' - ' para exibirmos de forma mais organizada em print() após a vírgula.

Em seguida, pegaremos a informação dentro de cada tag com info.get_text() ou info.getText().

In [100]:
for info in infos:
    print(info.get('class'), ' - ', info.get_text())

['txt-name', 'inline']  -  LAMBORGHINI AVENTADOR
['txt-category', 'badge', 'badge-secondary', 'inline']  -  USADO
['txt-motor']  -  Motor 1.8 16v
['txt-description']  -  Ano 1993 - 55.286 km
['txt-location']  -  Belo Horizonte - MG


Com isso, veremos as classes e seus conteúdos textuais separados pelo traço adicionado.

Na célula seguinte, selecionaremos somente o primeiro atributo da classe. Como se trata de uma lista, o primeiro item será o de índice 0.

In [101]:
for info in infos:
    print(info.get('class')[0], ' - ', info.get_text())

txt-name  -  LAMBORGHINI AVENTADOR
txt-category  -  USADO
txt-motor  -  Motor 1.8 16v
txt-description  -  Ano 1993 - 55.286 km
txt-location  -  Belo Horizonte - MG


Com isso, teremos apenas o nome do atributo e seu texto.

Para retirarmos o txt- e obtermos apenas name, category, motor, description e location, usaremos o .split() após o índice 0 da primeira posição da lista.

Se não passarmos nada para o método, este considerará que deverá pegar a partir da divisão feita pelo espaço em branco entre as palavras.

Como temos apenas um traço dividindo o que queremos do que não queremos, deveremos passar '-' para .split().

In [102]:
for info in infos:
    print(info.get('class')[0].split('-'), ' - ', info.get_text())

['txt', 'name']  -  LAMBORGHINI AVENTADOR
['txt', 'category']  -  USADO
['txt', 'motor']  -  Motor 1.8 16v
['txt', 'description']  -  Ano 1993 - 55.286 km
['txt', 'location']  -  Belo Horizonte - MG


Com a execução, criamos outro array.

Mas agora queremos apenas a última informação da lista, então deveremos adicionar o índice [-1] após split().

Ao invés de ficarmos apenas printando, apagaremos esse print() e o substituiremos por card[] que englobará o código, enquanto igualaremos tudo a info.get_text().

In [103]:
for info in infos:
    card[info.get('class')[0].split('-')[-1]] = info.get_text()

Com isso, teremos a variável name com o valor Lamborghini Aventador e as demais informações. Quando chamarmos o card que só tinha o valor do veículo antes, receberemos os demais dados capturados pela varredura: categoria, motor, descrição e localização.

Com este novo conjunto de informações da nossa página, falta-nos pegar a informação dos acessórios, o que veremos no próximo passo.

Resumindo, juntaremos este último código com a variável infos que criamos para efetivar a nossa operação.

### <font color=red>Resumo

In [104]:
# Informações

infos = anuncio.find('div', {'class': 'body-card'}).findAll('p')
for info in infos:
    card[info.get('class')[0].split('-')[-1]] = info.get_text()

In [105]:
infos

[<p class="txt-name inline">LAMBORGHINI AVENTADOR</p>,
 <p class="txt-category badge badge-secondary inline">USADO</p>,
 <p class="txt-motor">Motor 1.8 16v</p>,
 <p class="txt-description">Ano 1993 - 55.286 km</p>,
 <p class="txt-location">Belo Horizonte - MG</p>]

Com isso, conseguiremos copiar para fazer a iteração nos outros cards.

# 5.4. Obtendo os ACESSÓRIOS do veículo anunciado

Neste passo, a última etapa da coleta de informações do nosso primeiro card será da parte de acessórios.

Após o ano e a quilometragem, encontraremos quatro características do veículo descritas.

No passo anterior, vimos que tudo isso está dentro de uma **div** com classe "body-card", a qual tem várias tags **p** que já selecionamos, e também uma tag **ul** de classe "1st-items".

Abrindo o inspetor de código de nossa página com "F12", perceberemos que esta marcação cria uma lista dentro do HTML. Expandindo-a, encontraremos as tags filhas **li** que representam cada um dos itens da lista.

De volta ao notebook, selecionaremos essas informações escrevendo a variável anuncio com .find() recebendo a mesma 'div' selecionada no passo anterior. A 'class' será a 'body-card' e executaremos a célula na parte "5.4".

In [106]:
anuncio.find('div', {'class': 'body-card'})

<div class="col-md-6 body-card">
<p class="txt-name inline">LAMBORGHINI AVENTADOR</p>
<p class="txt-category badge badge-secondary inline">USADO</p>
<p class="txt-motor">Motor 1.8 16v</p>
<p class="txt-description">Ano 1993 - 55.286 km</p>
<ul class="lst-items">
<li class="txt-items">► 4 X 4</li>
<li class="txt-items">► Câmera de estacionamento</li>
<li class="txt-items">► Controle de tração</li>
<li class="txt-items">► Sensor de estacionamento</li>
<li class="txt-items">...</li>
</ul>
<p class="txt-location">Belo Horizonte - MG</p>
</div>

Como retorno, notaremos a tag **ul** que queremos selecionar.

Há duas formas de fazermos isso; a primeira é adicionarmos .find() recebendo 'ul', e a segunda é adicionarmos apenas .ul ao final.

In [107]:
anuncio.find('div', {'class': 'body-card'}).ul

<ul class="lst-items">
<li class="txt-items">► 4 X 4</li>
<li class="txt-items">► Câmera de estacionamento</li>
<li class="txt-items">► Controle de tração</li>
<li class="txt-items">► Sensor de estacionamento</li>
<li class="txt-items">...</li>
</ul>

Iremos pegar todos os cinco itens **li** internos que possuem os dados que nos interessam.

Adicionando .findAll() recebendo 'li', geraremos uma lista no Python. A partir disso, o trabalho será basicamente organizar e retirar excesso de informações como já vimos anteriormente.

In [108]:
anuncio.find('div', {'class': 'body-card'}).ul.findAll('li')

[<li class="txt-items">► 4 X 4</li>,
 <li class="txt-items">► Câmera de estacionamento</li>,
 <li class="txt-items">► Controle de tração</li>,
 <li class="txt-items">► Sensor de estacionamento</li>,
 <li class="txt-items">...</li>]

Na célula seguinte, armazenaremos este comando em uma nova variável chamada items, que é a tradução de "itens" para o inglês. Em seguida, apenas a chamaremos para exibirmos o resultado.

In [109]:
items = anuncio.find('div', {'class': 'body-card'}).ul.findAll('li')
items

[<li class="txt-items">► 4 X 4</li>,
 <li class="txt-items">► Câmera de estacionamento</li>,
 <li class="txt-items">► Controle de tração</li>,
 <li class="txt-items">► Sensor de estacionamento</li>,
 <li class="txt-items">...</li>]

Com isso, faremos uma varredura dentro desta variável.

Mas antes, retiraremos o último item da lista com .pop() para eliminarmos as reticências, pois não nos interessam.

In [111]:
items.pop()

<li class="txt-items">► Sensor de estacionamento</li>

Executando este comando, exibiremos o item que foi retirado. Chamando apenas items na célula seguinte, repararemos que a linha **li** com as reticências foi removida de fato.

Este será o primeiro tratamento. Como segundo, eliminaremos o caractere especial de setinha. Mas antes, teremos que varrer novamente porque os itens estão dentro de uma lista, então aplicaremos o .getText() para pegarmos apenas o conteúdo textual da tag.

Então, usaremos for para cada item dentro de items como já vimos anteriormente.

Dentro do corpo, imprimiremos o resultado com print() recebendo item.getText().

In [112]:
for item in items:
    print(item.getText())

► 4 X 4
► Câmera de estacionamento
► Controle de tração


Para eliminarmos o caractere especial de setinha, aplicaremos o método .replace() dentro de print().

Em seguida, copiaremos o caractere e colaremos dentro dos parênteses de .replace() entre aspas simples. Como segundo argumento, diremos que não queremos nada em seu lugar, ou seja, passaremos apenas as aspas ''.

In [113]:
for item in items:
    print(item.getText().replace('► ', ''))

4 X 4
Câmera de estacionamento
Controle de tração


Executando este comando, teremos eliminado a setinha.

Dando continuidade, faremos esta variável um jeito diferente, porque futuramente criaremos um DataFrame e um problema pode acontecer quando temos um valor único ou uma lista array no lugar de uma variável.

Então colocaremos todas essas informações dentro de uma única coluna e dentro de um único valor do nosso primeiro card, o qual é um dicionário.

Para isso, criaremos uma lista na variável chamada de acessorios. Em seguida, usaremos o mesmo código de for logo na sequência.

Mas não queremos mais esse print(), pois não mostraremos nada. Em seu lugar, pegaremos acessorios e popularemos uma lista Python com .append().

Com isso, o sistema lerá cada um dos itens e os colocará dentro da lista.

Por fim, chamaremos apenas acessorios para exibirmos o resultado.

In [114]:
acessorios = []
for item in items:
    acessorios.append(item.getText().replace('► ', ''))

acessorios

['4 X 4', 'Câmera de estacionamento', 'Controle de tração']

Com a lista acessorios contendo todas as informações, aplicaremos o mesmo processo feito anteriormente com card[] recebendo 'items' e sendo igual a acessorios para criarmos o card.

Na célula seguinte, chamaremos apenas card.

In [115]:
card['items'] = acessorios

In [116]:
card

{'value': 'R$ 338.000',
 'name': 'LAMBORGHINI AVENTADOR',
 'category': 'USADO',
 'motor': 'Motor 1.8 16v',
 'description': 'Ano 1993 - 55.286 km',
 'location': 'Belo Horizonte - MG',
 'items': ['4 X 4', 'Câmera de estacionamento', 'Controle de tração']}

Como retorno, veremos as informações de valor, nome, categoria, coisa e etc. Ao final, encontraremos a chave 'items' cujo valor é uma lista.

Em resumo, executamos o seguinte código:

### <font color=red>Resumo

In [117]:
# Acessórios

items = anuncio.find('div', {'class': 'body-card'}).ul.findAll('li')
items.pop()
acessorios = []
for item in items:
    acessorios.append(item.getText().replace('► ', ''))
card['items'] = acessorios

Quando formos criar o DataFrame, teremos um problema; porém, aprenderemos adiante uma maneira de usar o Pandas para resolvê-lo quando tivermos mais informações.

Com isso, criaremos um modelo de Regressão em cima desses dados de precificação de veículos para prevermos os valores com o conjunto de informações de um carro.

# 5.5 Criando um DataFrame com os dados coletados do Alura Motors

A biblioteca Pandas é muito importante para quem trabalha com Data Science, e veremos como usá-la para resolver alguns possíveis questões, apesar de não ser o foco deste treinamento.

Obtivemos um conjunto de informações do anúncio dentro da variável card, e precisaremos armazená-lo em um formato que nos permita realizar análises, criar Regressões e etc. Para isso, usaremos a estrutura de dados do Pandas por conta do DataFrame.

Na primeira célula da parte "5.5" de nosso notebook, a primeira tarefa será importar pandas como seu "apelido" padrão pd.

In [118]:
import pandas as pd

Para gerarmos um DataFrame, criaremos uma variável chamada dataset e a igualaremos ao pd com o método .DataFrame() recebendo card.

Em seguida, o exibiremos chamando apenas dataset na próxima célula.

In [120]:
dataset = pd.DataFrame(card)

In [121]:
dataset

value                   name category          motor  \
0  R$ 338.000  LAMBORGHINI AVENTADOR    USADO  Motor 1.8 16v   
1  R$ 338.000  LAMBORGHINI AVENTADOR    USADO  Motor 1.8 16v   
2  R$ 338.000  LAMBORGHINI AVENTADOR    USADO  Motor 1.8 16v   
3  R$ 338.000  LAMBORGHINI AVENTADOR    USADO  Motor 1.8 16v   

            description             location                     items  
0  Ano 1993 - 55.286 km  Belo Horizonte - MG                     4 X 4  
1  Ano 1993 - 55.286 km  Belo Horizonte - MG  Câmera de estacionamento  
2  Ano 1993 - 55.286 km  Belo Horizonte - MG        Controle de tração  
3  Ano 1993 - 55.286 km  Belo Horizonte - MG  Sensor de estacionamento

Porém, geramos um DataFrame com quatro linhas ao invés de uma única correspondente ao primeiro card.

Notaremos que a lista items está mal interpretada e as outras informações estão repetidas, pois queremos que cada anúncio diferente esteja em uma linha do DataFrame, e que os itens deste array estejam na mesma do carro a que pertencem.

Para resolvermos esse problema, alteraremos o dataset que está errado; no lugar de (card), inseriremos o método .from_dict() trazendo o card.

Passaremos uma orientação como segundo argumento chamada orient que não será por colunas que é o padrão, e sim por 'index' com os valores numéricos da primeira coluna que identificam as linhas.

In [122]:
dataset = pd.DataFrame.from_dict(card, orient = 'index')
dataset

0
value                                               R$ 338.000
name                                     LAMBORGHINI AVENTADOR
category                                                 USADO
motor                                            Motor 1.8 16v
description                               Ano 1993 - 55.286 km
location                                   Belo Horizonte - MG
items        [4 X 4, Câmera de estacionamento, Controle de ...

Também notaremos que a linha de items contém todos os elementos desta lista, como precisávamos.

Para organizarmos e transformarmos em uma tabela normal, bastará inserirmos simplesmente .T ao final do código. Com isso, iremos transpor a tabela e trocar a disposição das linhas por colunas.

In [123]:
dataset = pd.DataFrame.from_dict(card, orient = 'index').T
dataset

value                   name category          motor  \
0  R$ 338.000  LAMBORGHINI AVENTADOR    USADO  Motor 1.8 16v   

            description             location  \
0  Ano 1993 - 55.286 km  Belo Horizonte - MG   

                                               items  
0  [4 X 4, Câmera de estacionamento, Controle de ...

Com isso, poderemos criar uma tabela externa em um arquivo .csv, por meio do método .to_csv() aplicado no dataset.

Dentro dos parênteses, passaremos o endereço da pasta chamada "data" dentro do diretório "output" presente nos arquivos disponibilizados para este curso, e em seguida o nome do arquivo dataset.csv.

Como segundo argumento de .to_csv(), passaremos um separador sep sendo igual a ';'. Não queremos a identificação do index em nosso caso, então o igualaremos a False.

O quarto argumento encoding será 'utf-8-sig', pois estamos trabalhando com dados em inglês e português, e não queremos gerar confusão.

In [124]:
dataset.to_csv('./output/data/dataset.csv', sep=';', index = False, encoding = 'utf-8-sig')

In [125]:
dataset

value                   name category          motor  \
0  R$ 338.000  LAMBORGHINI AVENTADOR    USADO  Motor 1.8 16v   

            description             location  \
0  Ano 1993 - 55.286 km  Belo Horizonte - MG   

                                               items  
0  [4 X 4, Câmera de estacionamento, Controle de ...

Executando esta célula, encontraremos o arquivo dataset.csv na pasta em que foi salvo.

Abrindo-o, acessaremos as informações que capturamos e também poderemos utilizá-lo em outro lugar ou fazer um import para dentro do Python mesmo.

A seguir, começaremos a capturar as imagens do Alura Motors para salvarmos e depois fazermos alguns outros modelos e análises em dataset.

# 5.6. Obtendo a FOTO do anúncio

Com quase tudo pronto no nosso procedimento do Scraping do primeiro card do Alura Motors, falta-nos somente a obtenção da imagem.

Como cientistas de dados, poderemos ter de criar um conteúdo de Machine Learning que faça reconhecimento de objetos em imagens.

Para isso, passaremos um conjunto de dados com muitas imagens contendo o objeto que nos interessa para treinarmos esse modelo, como identificar veículos em fotografias por exemplo.

Nisso o Web Scraping nos ajudará bastante a coletar uma quantidade muito grande de dados de forma bem rápida, para então realizarmos o nosso projeto.

Primeiro, identificamos o "well card" como o primeiro card, o qual possui três **div**; a primeira é onde está a imagem com class de "image-card". Abrindo-a no inspetor de código do site Alura Motors, encontraremos uma tag **img** que mostra imagens.

Passando o cursor do mouse em cima desta parte em "Elements", destacaremos o atributo com informação que armazena a figura que queremos.

De volta ao nosso notebook, escreveremos anuncio.find() recebendo a 'div' como já fizemos anteriormente. A class será a image-card desta vez.

In [126]:
anuncio.find('div', {'class': 'image-card'})

<div class="col-md-3 image-card">
<img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg" width="220"/>
</div>

Notaremos que a informação que queremos está dentro de um atributo chamado src, que é a fonte dos dados. Vamos selecionar apenas a tag <img> com .img. Armazenaremos este comando na variável image e a exibiremos.

In [127]:
image = anuncio.find('div', {'class': 'image-card'}).img
image

<img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg" width="220"/>

Para obtermos o valor de um atributo dentro de uma tag, passaremos o image.get() recebendo 'src'.

In [128]:
image.get('src')

'https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg'

Ao executarmos, receberemos o caminho que exibe a imagem no site. Se imprimirmos este código com print(), receberemos o link clicável que nos direciona para a figura.

Como um item extra, adicionamos uma forma de exibirmos as informações buscadas no HTML diretamente no notebook, caso seja útil para estudos futuros.

### Visualizando a FOTO no notebook (extra)

In [129]:
from IPython.core.display import display, HTML

display(HTML(str(anuncio.find('div', {'class': 'image-card'}).img)))

Se não passarmos o link direto teremos um erro, então é interessante transformarmos em string para exibirmos a fotografia direto no notebook.

Como temos uma grande quantidade de imagens, seria muito trabalhoso fazer o download de todas manualmente, então o faremos automaticamente.

Para isso, temos a funcionalidade urlretrieve dentro do urllib.request que já conhecemos, cuja documentação pode ser acessada neste link.

Passaremos a url que acabamos de ver e o lugar onde queremos fazer o download.

### Rotina para acessar e salvar a FOTO do anúncio

## https://docs.python.org/3/library/urllib.request.html#urllib.request.urlretrieve

In [130]:
from urllib.request import urlretrieve

Dentro de urlretrieve(), passaremos o image.get() recebendo 'src' e o o endereço da pasta "img" dentro de "output" onde queremos fazer o download.

Precisaremos passar o nome, e para isso escreveremos um outro / e concatenaremos com a informação do próprio name dado ao carro.

Para isso, escreveremos na célula em branco anterior para aprendermos um atalho.

Chamaremos somente image.get() recebendo a fonte, e aplicaremos .split() recebendo a barra '/' como divisória. No último item da lista, teremos o nome do arquivo que queremos, e o acessaremos passando o índice -1 dentro de colchetes.

In [131]:
image.get('src').split('/')[-1]

'lamborghini-aventador-2932196__340.jpg'

Na célula seguinte com o código que havíamos começado, faremos a concatenação disso com o caminho da imagem em urilretrieve(), desta forma:

### <font color=red>Resumo

In [132]:
from urllib.request import urlretrieve

urlretrieve(image.get('src'), './output/img/' + image.get('src').split('/')[-1])

('./output/img/lamborghini-aventador-2932196__340.jpg',
 <http.client.HTTPMessage at 0x1c6262b01f0>)

Rodando este bloco, poderemos encontrar o arquivo de imagem na pasta "img", o qual será usado no modelo de Machine Learning.

A seguir, juntaremos os códigos para acessarmos todos os 10 itens que estão nesta primeira página, e depois o site inteiro.



# <font color=green>6. WEB SCRAPING DO SITE ALURA MOTORS - OBTENDO OS DADOS DE TODOS OS ANÚNCIOS DE UMA PÁGINA
    


# 6.1. Identificando as informações no HTML

Já conseguimos obter todas as informações do primeiro card do Alura Motors.

Neste passo, descobriremos uma forma de aplicar esse mesmo procedimento para todos os outros e obter todos os dados dos 10 anúncios da primeira página.

Sabemos como iterar por esses cards; observando a aba "Elements" do inspetor de código do Alura Motors no navegador, cada anúncio está dentro de uma **div class="well card"** como vimos nos passos anteriores.

Porém, todas essas **div** estão dentro de uma outra cujo id é igual a "container-cards", a qual seria a parent ou "pai".

O id é também um atributo que vem do arquivo CSS e que indica o tipo de formatação que aplicaremos nessa tag, mas este é único e não se repete ao longo do HTML, enquanto a classe pode aparecer igual mais de uma vez.

A class="well card" está repetindo para várias marcações enquanto o id="container-cards" é o único no documento HTML. Então o acessaremos e veremos como criar uma lista para podermos iterar sobre esta.

O ideal é que criemos uma lista com informações de cada card como sendo os itens da lista, o que é bem fácil de fazer. Como não estamos mais trabalhando com aquele anuncio que criamos, voltaremos para o soup já carregado na primeira célula da parte "6.1" de nosso notebook, o qual contém toda a informação HTML da pasta.

Começaremos acessando a 'div' com soup.find(). Após a vírgula, passaremos um dicionário Python com {} onde terá o "id" cujo conteúdo é "container-cards".

In [133]:
soup.find('div', {"id": "container-cards"})

<div id="container-cards" style="height: 100%">
<div class="well card">
<div class="col-md-3 image-card">
<img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg" width="220"/>
</div>
<div class="col-md-6 body-card">
<p class="txt-name inline">LAMBORGHINI AVENTADOR</p>
<p class="txt-category badge badge-secondary inline">USADO</p>
<p class="txt-motor">Motor 1.8 16v</p>
<p class="txt-description">Ano 1993 - 55.286 km</p>
<ul class="lst-items">
<li class="txt-items">► 4 X 4</li>
<li class="txt-items">► Câmera de estacionamento</li>
<li class="txt-items">► Controle de tração</li>
<li class="txt-items">► Sensor de estacionamento</li>
<li class="txt-items">...</li>
</ul>
<p class="txt-location">Belo Horizonte - MG</p>
</div>
<div class="col-md-3 value-card">
<div class="value">
<p class="txt-value">R$ 338.000</p>
</div>
</div>
</div>
<div class="well card">
<div class="col-md-3 ima

Desta forma, já conseguiremos acessar este atributo inteiro.

Sabemos que há diversas **div** dentro deste elemento cuja class é igual a "well card". Para acessá-las, criaremos uma lista e colocaremos cada tag como um item.

Usaremos o método .findAll() recebendo a 'div' que já conhecemos. Usaremos a outra forma de acessá-la com class_, e poderemos igualar a "well" ou "card". Escolheremos o segundo para ficar mais simples de entendermos.

In [134]:
soup.find('div', {"id": "container-cards"}).findAll('div', class_="card")

[<div class="well card">
 <div class="col-md-3 image-card">
 <img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg" width="220"/>
 </div>
 <div class="col-md-6 body-card">
 <p class="txt-name inline">LAMBORGHINI AVENTADOR</p>
 <p class="txt-category badge badge-secondary inline">USADO</p>
 <p class="txt-motor">Motor 1.8 16v</p>
 <p class="txt-description">Ano 1993 - 55.286 km</p>
 <ul class="lst-items">
 <li class="txt-items">► 4 X 4</li>
 <li class="txt-items">► Câmera de estacionamento</li>
 <li class="txt-items">► Controle de tração</li>
 <li class="txt-items">► Sensor de estacionamento</li>
 <li class="txt-items">...</li>
 </ul>
 <p class="txt-location">Belo Horizonte - MG</p>
 </div>
 <div class="col-md-3 value-card">
 <div class="value">
 <p class="txt-value">R$ 338.000</p>
 </div>
 </div>
 </div>,
 <div class="well card">
 <div class="col-md-3 image-card">
 <img alt="

Com a execução desta célula, criamos uma lista que engloba entre colchetes os diversos elementos do bloco separados por vírgula. Cada item possui a informação de um card, e é isso que precisaremos para fazermos a iteração.

Em seguida, confirmaremos que são realmente 10 anúncios com o método len() recebendo o código da lista que criamos.

In [135]:
len(soup.find('div', {"id": "container-cards"}).findAll('div', class_="card"))

10

Com isso, perceberemos os 10 itens dentro da lista.

Armazenaremos essas informações dos cards em uma variável chamada anuncios na célula seguinte.

In [136]:
anuncios = soup.find('div', {"id": "container-cards"}).findAll('div', class_="card")

Para fazermos a iteração na próxima célula, digitaremos for para cada anuncio, dentro de anuncios com in. Em seguida, os exibiremos com print() e transformaremos o anuncio de um objeto BeautifulSoup em string com o método str().

Isso nos permitirá concatenar com duas linhas "\n\n", ou seja, pular duas linhas para ficar claro onde começa e onde termina.

In [137]:
for anuncio in anuncios:
    print(str(anuncio) + "\n\n")

<div class="well card">
<div class="col-md-3 image-card">
<img alt="Foto" height="155" src="https://caelum-online-public.s3.amazonaws.com/1381-scraping/01/img-cars/lamborghini-aventador/lamborghini-aventador-2932196__340.jpg" width="220"/>
</div>
<div class="col-md-6 body-card">
<p class="txt-name inline">LAMBORGHINI AVENTADOR</p>
<p class="txt-category badge badge-secondary inline">USADO</p>
<p class="txt-motor">Motor 1.8 16v</p>
<p class="txt-description">Ano 1993 - 55.286 km</p>
<ul class="lst-items">
<li class="txt-items">► 4 X 4</li>
<li class="txt-items">► Câmera de estacionamento</li>
<li class="txt-items">► Controle de tração</li>
<li class="txt-items">► Sensor de estacionamento</li>
<li class="txt-items">...</li>
</ul>
<p class="txt-location">Belo Horizonte - MG</p>
</div>
<div class="col-md-3 value-card">
<div class="value">
<p class="txt-value">R$ 338.000</p>
</div>
</div>
</div>


<div class="well card">
<div class="col-md-3 image-card">
<img alt="Foto" height="155" src="ht

Como retorno da execução, veremos as informações dos cards separados por duas linhas, facilitando a iteração.

A seguir, juntaremos tudo o que fizemos nos passos anteriores em um único bloco de código para obtermos os dados dos 10 cards de uma vez só.

# 6.2. Criando uma rotina de scraping

Já conseguimos acessar as informações de cada um dos dez anúncios de veículos exibidos na primeira página do site Alura Motors.

A partir de agora, aplicaremos todo o conhecimento adquirido e o que construímos nos passos anteriores dentro de um iterador, o qual varrerá todas as informações destes cards.

Encontraremos duas células com um resumo de códigos criados para obter os dados até então na parte "6.2" do nosso notebook, bem como as importações de bibliotecas, alguns códigos prontos e comentários para preenchermos neste passo:

In [138]:
# Valor
card['value'] = anuncio.find('p', {'class': 'txt-value'}).getText()

# Informações
infos = anuncio.find('div', {'class': 'body-card'}).findAll('p')
for info in infos:
    card[info.get('class')[0].split('-')[-1]] = info.get_text()

# Acessórios
items = anuncio.find('div', {'class': 'body-card'}).ul.findAll('li')
items.pop()
acessorios = []
for item in items:
    acessorios.append(item.get_text().replace('► ', ''))
card['items'] = acessorios

# Imagens
image = anuncio.find('div', {'class': 'image-card'}).img
urlretrieve(image.get('src'), './output/img/' + image.get('src').split('/')[-1])     

('./output/img/toyota-3245397__340.jpg',
 <http.client.HTTPMessage at 0x1c6262c5790>)

In [143]:
# Importando bibliotecas
from urllib.request import urlopen, urlretrieve
from bs4 import BeautifulSoup
import pandas as pd

# Declarando variável cards

cards = []

# Obtendo o HTML
response = urlopen('https://alura-site-scraping.herokuapp.com/index.php')
html = response.read().decode('utf-8')
soup = BeautifulSoup(html, 'html.parser')

# Obtendo as TAGs de interesse
anuncios = soup.find('div', {"id": "container-cards"}).findAll('div', class_="card")


# Coletando as informações dos cards
for anuncio in anuncios:
    card = {}

    # Valor
    card['value'] = anuncio.find('p', {'class': 'txt-value'}).getText()

    # Informações
    infos = anuncio.find('div', {'class': 'body-card'}).findAll('p')
    for info in infos:
        card[info.get('class')[0].split('-')[-1]] = info.get_text()

    # Acessórios
    items = anuncio.find('div', {'class': 'body-card'}).ul.findAll('li')
    items.pop()
    acessorios = []
    for item in items:
        acessorios.append(item.get_text().replace('► ', ''))
    card['items'] = acessorios
    
    # Adicionando resultado à lista cards
    cards.append(card)

    # Imagens
    image = anuncio.find('div', {'class': 'image-card'}).img
    urlretrieve(image.get('src'), './output/img/' + image.get('src').split('/')[-1])      
    

# Criando um DataFrame com os resultados
dataset = pd.DataFrame(cards)
dataset.to_csv('./output/data/dataset.csv', sep=';', index = False, encoding = 'utf-8-sig')
dataset

value                    name category            motor  \
0  R$ 338.000   LAMBORGHINI AVENTADOR    USADO    Motor 1.8 16v   
1  R$ 346.000                  BMW M2    USADO    Motor 3.0 32v   
2  R$ 480.000                    ALFA    USADO    Motor 1.8 16v   
3  R$ 133.000                   PUECH    USADO  Motor Diesel V8   
4  R$ 175.000  LAMBORGHINI MURCIELAGO    USADO     Motor 1.0 8v   
5  R$ 239.000            ASTON MARTIN    USADO  Motor Diesel V6   
6  R$ 115.000                     TVR    USADO  Motor 4.0 Turbo   
7  R$ 114.000               EXCALIBUR    USADO    Motor 3.0 32v   
8   R$ 75.000                 MCLAREN     NOVO     Motor Diesel   
9  R$ 117.000                  TOYOTA    USADO  Motor 4.0 Turbo   

            description             location  \
0  Ano 1993 - 55.286 km  Belo Horizonte - MG   
1  Ano 2018 - 83.447 km  Belo Horizonte - MG   
2  Ano 2004 - 19.722 km  Rio de Janeiro - RJ   
3  Ano 1992 - 34.335 km       São Paulo - SP   
4     Ano 1991 - 464 km  Belo Horizonte - MG   
5  Ano 2004 - 50.189 km  Belo Horizonte - MG   
6  Ano 2014 - 17.778 km  Belo Horizonte - MG   
7  Ano 2009 - 81.251 km  Rio de Janeiro - RJ   
8       Ano 2019 - 0 km       São Paulo - SP   
9  Ano 1999 - 12.536 km       São Paulo - SP   

                                               items   opportunity  
0  [4 X 4, Câmera de estacionamento, Controle de ...           NaN  
1  [Câmera de estacionamento, Controle de estabil...           NaN  
2  [Central multimídia, Bancos de couro, Rodas de...           NaN  
3  [Bancos de couro, Freios ABS, Rodas de liga, C...           NaN  
4  [Central multimídia, Teto panorâmico, Sensor c...           NaN  
5  [Painel digital, Controle de tração, Teto pano...  OPORTUNIDADE  
6  [4 X 4, Teto panorâmico, Central multimídia, C...           NaN  
7  [Painel digital, Câmbio automático, Sensor de ...           NaN  
8  [Central multimídia, Câmera de estacionamento,...           NaN  
9  [Bancos de couro, Freios ABS, Piloto automátic...  OPORTUNIDADE

Notaremos a criação de uma nova variável chamada opportunity, porque em alguns casos encontramos a palavra "usado" e "oportunidade" que provavelmente indica um veículo em promoção.

Tivemos sorte desta informação estar dentro uma tag <p> cuja classe é igual a "txt-opportunity" dentro da <div> com classe body-card, pois o código já a identificou e criou a variável de forma simples.

Quando não possui informação alguma nesta parte, notaremos o valor nulo NaN, mas depois faremos um tratamento com Pandas.

A seguir, basicamente faremos a mesma coisa para acessar todas as informações dos 246 veículos totais do site de uma só vez.

# <font color=green>7. WEB SCRAPING DO SITE ALURA MOTORS - OBTENDO OS DADOS DE TODOS OS ANÚNCIOS DO SITE

# 7.1. Identificando as informações no HTML

Nesta última etapa, faremos o Scraping da página inteira.

Notaremos que 246 veículos foram encontrados ao total, distribuídos em 25 páginas com informações. Como já obtivemos os dados dos anúncios da primeira página, passaremos pelas demais coletando automaticamente.

Quando passarmos para a segunda página, repararemos uma modificação na url da Alura Motors ao final do endereço; o símbolo de interrogação ? passa um parâmetro chamado page=2, e se formos para a terceira mudará para page=3 e assim sucessivamente até última.

Inclusive, poderemos alterar diretamente na barra de busca se quisermos navegar pelas páginas.

A de número "25" já não tem 10 cards, e sim apenas 6 informações. Antes de continuarmos, conferiremos se conseguimos voltar à página inicial escrevendo page=1 ao final da url.

Se sim, conseguiremos varrer cada uma dessas páginas somente mudando esse parâmetro, mas queremos fazer isso de forma automática.

Como é um site de anúncios, as pessoas anunciam ou retiram o card constantemente, mas em geral possui uma tendência a aumentar. Portanto, não teremos 25 páginas para sempre, pois esse número pode se alterar conforme o número total de cards no site, e por isso queremos que o código o faça automaticamente.

No Alura Motors aberto no navegador, clicaremos com o botão direito do mouse sobre o número "25" que indica o total de páginas e selecionaremos a opção "Inspecionar". Na aba "Elements", veremos que se trata de uma string dentro de uma tag <span>, o qual é um marcador que tem uma classe chamada "info-pages".

Acessaremos de forma direta, pois já conhecemos o BeautifulSoup e os métodos para fazê-lo.

Na parte "7.1" de nosso notebook, escreveremos soup.find() recebendo 'span' e a class_ sem nos esquecermos do underscore _ igual a "info-pages".

In [145]:
soup.find('span', class_="info-pages")

<span class="info-pages">Página 1 de 25</span>

Com isso, teremos a informação desta tag, mas queremos obter somente o número "25".

Primeiro, aplicaremos o .get_text() ao final do código copiado na célula seguinte para fazermos parte a parte.

In [146]:
soup.find('span', class_="info-pages").get_text()

'Página 1 de 25'

Em seguida, usaremos o método .split() sem nada dentro para partirmos o texto nos espaços em branco.

In [147]:
soup.find('span', class_="info-pages").get_text().split()

['Página', '1', 'de', '25']

Na lista criada, o número '25' será o último item. Para o pegarmos, adicionaremos o índice [-1].

In [148]:
soup.find('span', class_="info-pages").get_text().split()[-1]

'25'

Feito isso, receberemos a string com o valor '25', como queríamos.

Criaremos um iterador, então precisaremos de um número que vá de "1" até "25", por exemplo.

Para tal, bastará escrevermos int() recebendo todo o código anterior para transformá-lo em um número inteiro, pois não teremos valores com casas decimais.

In [149]:
int(soup.find('span', class_="info-pages").get_text().split()[-1])

25

Agora, além de termos a iteração na primeira página, passaremos por cada uma das demais e faremos a varredura dentro delas todas as vezes que forem necessárias, como veremos a seguir.

# 7.2. Criando uma rotina de scraping

Para finalizarmos, faremos uma raspagem de todo o site Alura Motors; pegaremos todas as informações e as armazenaremos para realizarmos nossos modelos de Machine Learning e Data Science posteriormente.

Como percebemos no passo anterior, o endereço url sofre algumas modificações conforme avançamos de página.

Na primeira célula da parte "7.2" do notebook, copiaremos tudo o que fizemos na parte "6.2", ou seja, o código da varredura de todas as informações de uma página.

Também haverá algumas modificações e comentários novos para preenchermos neste último passo:

In [19]:
# Importando bibliotecas
from urllib.request import urlopen, urlretrieve
from bs4 import BeautifulSoup
import pandas as pd

# Declarando variável cards
cards = []

## Obtendo o HTML e o total de páginas
response = urlopen('https://alura-site-scraping.herokuapp.com/index.php')
html = response.read().decode('utf-8')
soup = BeautifulSoup(html, 'html.parser')
pages = int(soup.find('span', class_="info-pages").get_text().split()[-1])

## Iterando por todas as páginas do site
for i in range(pages):
    ## Obtendo o HTML
    response = urlopen('https://alura-site-scraping.herokuapp.com/index.php?page=' + str(i + 1))
    html = response.read().decode('utf-8')
    soup = BeautifulSoup(html, 'html.parser')

    # Obtendo as TAGs de interesse
    anuncios = soup.find('div', {"id": "container-cards"}).findAll('div', class_="card")

    # Coletando as informações dos CARDS
    for anuncio in anuncios:
        card = {}

        # Valor
        card['value'] = anuncio.find('p', {'class': 'txt-value'}).getText()

        # Informações
        infos = anuncio.find('div', {'class': 'body-card'}).findAll('p')
        for info in infos:
            card[info.get('class')[0].split('-')[-1]] = info.get_text()

        # Acessórios
        items = anuncio.find('div', {'class': 'body-card'}).ul.findAll('li')
        items.pop()
        acessorios = []
        for item in items:
            acessorios.append(item.get_text().replace('► ', ''))
        card['items'] = acessorios

        # Adicionando resultado a lista cards
        cards.append(card)

        # Imagens
        image = anuncio.find('div', {'class': 'image-card'}).img
        urlretrieve(image.get('src'), './output/img/' + image.get('src').split('/')[-1])     


# Criando um DataFrame com os resultados
dataset = pd.DataFrame(cards)
dataset.to_csv('./output/data/dataset.csv', sep=';', index = False, encoding = 'utf-8-sig')
dataset

value                    name category            motor  \
0    R$ 338.000   LAMBORGHINI AVENTADOR    USADO    Motor 1.8 16v   
1    R$ 346.000                  BMW M2    USADO    Motor 3.0 32v   
2    R$ 480.000                    ALFA    USADO    Motor 1.8 16v   
3    R$ 133.000                   PUECH    USADO  Motor Diesel V8   
4    R$ 175.000  LAMBORGHINI MURCIELAGO    USADO     Motor 1.0 8v   
..          ...                     ...      ...              ...   
241  R$ 489.000           SUV REAR TIRE    USADO    Motor 3.0 32v   
242  R$ 427.000                 ANTIQUE     NOVO    Motor 2.0 16v   
243  R$ 203.000                   SPORT    USADO    Motor 2.0 16v   
244  R$ 474.000                IMPERIAL    USADO    Motor 1.8 16v   
245  R$ 366.000          KIA SPORTS CAR    USADO    Motor 3.0 32v   

               description             location  \
0     Ano 1993 - 55.286 km  Belo Horizonte - MG   
1     Ano 2018 - 83.447 km  Belo Horizonte - MG   
2     Ano 2004 - 19.722 km  Rio de Janeiro - RJ   
3     Ano 1992 - 34.335 km       São Paulo - SP   
4        Ano 1991 - 464 km  Belo Horizonte - MG   
..                     ...                  ...   
241   Ano 1998 - 74.292 km       São Paulo - SP   
242        Ano 2019 - 0 km  Belo Horizonte - MG   
243  Ano 2001 - 102.776 km  Belo Horizonte - MG   
244  Ano 2011 - 101.787 km  Belo Horizonte - MG   
245   Ano 2001 - 88.564 km  Belo Horizonte - MG   

                                                 items   opportunity  
0    [4 X 4, Câmera de estacionamento, Controle de ...           NaN  
1    [Câmera de estacionamento, Controle de estabil...           NaN  
2    [Central multimídia, Bancos de couro, Rodas de...           NaN  
3    [Bancos de couro, Freios ABS, Rodas de liga, C...           NaN  
4    [Central multimídia, Teto panorâmico, Sensor c...           NaN  
..                                                 ...           ...  
241  [Câmera de estacionamento, Rodas de liga, Sens...           NaN  
242  [Bancos de couro, Freios ABS, Sensor de estaci...           NaN  
243  [Sensor crepuscular, Sensor de chuva, Vidros e...           NaN  
244  [Painel digital, Travas elétricas, Sensor de c...  OPORTUNIDADE  
245  [Sensor crepuscular, Bancos de couro, Sensor d...           NaN  

[246 rows x 8 columns]

In [20]:
dataset.head()

value                    name category            motor  \
0  R$ 338.000   LAMBORGHINI AVENTADOR    USADO    Motor 1.8 16v   
1  R$ 346.000                  BMW M2    USADO    Motor 3.0 32v   
2  R$ 480.000                    ALFA    USADO    Motor 1.8 16v   
3  R$ 133.000                   PUECH    USADO  Motor Diesel V8   
4  R$ 175.000  LAMBORGHINI MURCIELAGO    USADO     Motor 1.0 8v   

            description             location  \
0  Ano 1993 - 55.286 km  Belo Horizonte - MG   
1  Ano 2018 - 83.447 km  Belo Horizonte - MG   
2  Ano 2004 - 19.722 km  Rio de Janeiro - RJ   
3  Ano 1992 - 34.335 km       São Paulo - SP   
4     Ano 1991 - 464 km  Belo Horizonte - MG   

                                               items opportunity  
0  [4 X 4, Câmera de estacionamento, Controle de ...         NaN  
1  [Câmera de estacionamento, Controle de estabil...         NaN  
2  [Central multimídia, Bancos de couro, Rodas de...         NaN  
3  [Bancos de couro, Freios ABS, Rodas de liga, C...         NaN  
4  [Central multimídia, Teto panorâmico, Sensor c...         NaN